MaxText/inference/paged_attention.py

```diff
@@ -125,26 +126,67 @@ def paged_dot_product_attention_with_max_and_sum(self, query, key, value):
 
     return attn, local_max, local_sums
 
-  def paged_attention(
+  # TODO(rupliu): add sharding when SPMD is fully supported
+  def paged_attention_v2_prefill(
       self,
       query: Array,
       key_pages_var: nn.Variable,
       value_pages_var: nn.Variable,
       page_state: page_manager.PageState,
   ) -> Array:
-    """Apply Paged Attention.
-
-    Annotations:
-    b: batch_size
-    s: sequence length
-    n: query heads
+    """Apply ragged input Paged Attention in prefill only."""
+    k_p = jnp.permute_dims(key_pages_var.value, (1, 2, 0, 3))
+    v_p = jnp.permute_dims(value_pages_var.value, (1, 2, 0, 3))
+    c_q_l = jnp.array([0, page_state.sequence_lengths[0]])
+    num_seqs = jnp.array([1])
+    query = query[0]  # [batch_size, max_num_tokens, num_kv_heads, head_dim] to [max_num_tokens, num_kv_heads, head_dim]
+    result = paged_attention_kernel_v2.ragged_paged_attention(
+        q=query,
+        k_pages=k_p,  # [total_num_pages, page_size, num_kv_heads, head_dim]
+        v_pages=v_p,
+        kv_lens=jnp.array([query.shape[0]]),  # max_prefill_length
+        cu_q_lens=c_q_l,
```

Comment:
What exactly are c_q_l and cu_q_lens? It's not very clear just from the variable naming.

Reply:
SG! will update this